social_core/backends/discourse.py

```diff
@@ -0,0 +1,86 @@
+from .base import BaseAuth
+from ..exceptions import AuthException, AuthTokenError
+import hmac
+from base64 import b64encode, b64decode
+from hashlib import sha256
+import urllib
+from ..utils import parse_qs
+import time
+
+
+class DiscourseAuth(BaseAuth):
+    name = "discourse"
+    EXTRA_DATA = ["username", "name", "avatar_url"]
+
+    def auth_url(self):
+        """Return redirect url"""
+        returnUrl = self.redirect_uri
+        nonce = self.strategy.random_string(64)
+        self.add_nonce(nonce)
+
+        payload = "nonce=" + nonce + "&return_sso_url=" + returnUrl
+        base64Payload = b64encode(payload)
+        payloadSignature = hmac.new(
+            self.setting("SOCIAL_AUTH_DISCOURSE_AUTH_SECRET"), base64Payload, sha256
+        ).hexdigest()
+        encodedParams = urllib.urlencode(
+            {"sso": base64Payload, "sig": payloadSignature}
+        )
+        return (
+            self.setting("SOCIAL_AUTH_DISCOURSE_AUTH_SERVER_URL")
+            + "/session/sso_provider?"
+            + encodedParams
+        )
+
+    def get_user_id(self, details, response):
+        return response["email"]
+
+    def get_user_details(self, response):
+        results = {
+            "username": response.get("username"),
+            "email": response.get("email"),
+            "name": response.get("name"),
+            "groups": response.get("groups", "").split(","),
+            "is_staff": response.get("admin") == "true"
+            or response.get("moderator") == "true",
+            "is_superuser": response.get("admin") == "true",
+        }
+        return results
+
+    def add_nonce(self, nonce):
+        self.strategy.storage.nonce.use(
+            self.setting("SOCIAL_AUTH_DISCOURSE_AUTH_SERVER_URL"), time.time(), nonce
+        )
+
+    def get_nonce(self, nonce):
+        try:
+            return self.strategy.storage.nonce.objects.get(
```

Comment:
It seems like `nonce.use` is the only method available on the NonceMixin: https://github.com/python-social-auth/social-core/blob/cfb4fc028051e25e5766d7dfd479a35af247b4a7/social_core/storage.py#L199-L208

I used `nonce.use` above when creating the nonce, but I'm trying to retrieve the nonce on this line. Since I didn't see an API for this, I resorted to using the Django queryset api `nonce.objects`. I'm not sure how this will interact with non-Django strategies.

I also ran into a similar issue below, where I am deleting the Nonce by invoking a `.delete()` method on the retrieved Nonce model. The `.delete()` method is part of Django's ModelBase api - so I'm not sure whether this API is consistent on other strategies.

Reply:
As you point out, this imposes a dependency on Django that's not desired, the right way to do it is:

1. Define the non-implemented methods `get_nonce(...)` and `delete_nonce()` on the `NonceMixin` class
2. Then on each strategy (django, flask, etc) implement the corresponding methods